Functional Expansion module

BruceWyxv opened the pull request:
This is a submission for adding functional expansions to MOOSE.

The idea is to include functional expansions as a module. This keeps all the associated functionality in a common location, simplifying future development. It also avoids the addition of a new batch of code to the central framework. Not everyone will be using this functionality, so it seems to make sense to include it as a module.

Examples of use, tests, and unit tests are all included.

A few clean-up edits of modules/combine/src/base/CombinedApp.C and modules/modules.mk are included with the new additions.

closes #10577

Commit:
Initial Creation of Module

Followed instructions at http://mooseframework.org/wiki/CreatingANewMooseModule/
Cherry-picked autogen files from running `scripts/stork.sh FunctionalExpansionTools --module`

Commit:
Ran `git clang-format` to ensure formatting

Ignored changes in CompositSeriesBasisInterface.C:153-209 to keep whitespaces alignments

Commit:
Groomed code for submission to MOOSE

Commit:
Adding unit test files

Commit:
Finished unit testing files

Commit:
Updated Hashing and added unit test

Commit:
Reorganized existing tests

Commit:
Fixes

Updated mispelling in CombinedApp.C
Renamed input parameter to be physically correct

Commit:
Added 2D interface examples

Commit:
Updates to interface files

Commit:
Added volumetric value coupling examples

Commit:
Added README for module

Commit:
Minor user-specific text fixes in BC classes

Commit:
Added direct coupling example

Commit:
Updated README files

Added more TODO items
Added README for examples folder
Added more examples
Minor spelling fixes

Commit:
Fixed some formatting in FunctionSeries.C

Commit:
Completed all error files

Commit:
Updated parameter description in FunctionSeries.C

Commit:
Completed test files

Commit:
Slight README update

Commit:
Finished README.md

Commit:
Added ignore section for clang-format

Commit:
Fixed whitespace in modules.mk

permcody commented:
Thanks for posting this. It's a huge chunk of work to go through and there are several steps ahead to get this passing tests and mergeable. I think it's big enough that you should plan to come over to the MOOSE cube and go through the design with us. You could walk us through the basic idea and we can comment on any larger scale design issues that arise. In the meantime, you are welcome to keep making changes to see if you can get through the prechecks so that you can start testing this.

Let us know what your schedule looks like and when you could come over. No rush.

permcody commented:
@rcm59 - Large contribution coming through into the modules.

BruceWyxv commented:
That sounds like a great plan. I am working on the precheck issues as we speak.

I can come by tomorrow or Wednesday afternoon if either of those times work.

Commit:
Removed non-ASCII characters

Commit:
Replaced "cout" statements in MutableCoefficientsInterface.C

Added test to verify use
Closes #10577

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/147983/) on b03aa6f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10579/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10579/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7a18e52d8770a53d23ff4b961d11e9f0fdcc0dc6`

Commit:
Fixed clang-format issues

andrsd commented on modules/functional_expansion_tools/src/utils/MutableCoefficientsInterface.C:
```diff
@@ -0,0 +1,220 @@
+  _coefficients = other._coefficients;
+
+  if (_print_coefficients)
+    dynamic_cast<ConsoleStreamInterface *>(this)->_console << *this;
```
This looks like an instant segfault. `MutableCoefficientsInterface` does not inherit from `ConsoleStreamInterface`, so the result of you cast will be `nullptr` and doing `->` on it is bad.

BruceWyxv replied:
All the classes that use this interface (Function and UserObject, currently) do inherit from `ConsoleStreamInterface`. It is only intended to be used with these, and I am avoiding a diamond-inheritance scheme.

If working code is a concern, I can verify that it does work. I created the test `multiapp_print_coefficients.i` to ensure this works in the future as well.

If there is a design issue with this then I will gladly work with the team to come up with an alternative.

andrsd replied:
Also, I am not convinced about the `_print_coefficient` thing.  I'd expect you provide a getter, so people can get the coefficients and then they can print them using the regular `_console << data` syntax.

BruceWyxv replied:
I have no problem providing a `Stream` object containing a formatted table of the coefficients. Is this what you are referring to?

I already provide access to the raw data as a vector via the `GetCoefficients()` method. The input option `print_coefficients` is available for the users who are using the as-built module, not developing their own extensions.

BruceWyxv replied:
A getter `std::string getCoefficientsTable()` was just added.

andrsd replied:
There are multiple ways how to go about printing. I was suggesting that this class would only provide the raw data and the printing would be done by the caller. But, then I saw you overloaded `operator <<` and have some pretty printing there. So, I think you probably have a good reason to do so. My guess is that you do are reusing this class and do not want to replicate the code.

Providing a `std::string` through a function would be another good way.

Another solution to this problem is to pass the pointer to `MooseObject`-derived class that is using this interface and use that to get `_console` member, see `SetupInterface` for an example of what I am talking about.  Actually, that should be probably the way you should pick, so that your interface is consistent with the other interfaces in the framework.

BruceWyxv replied:
Exactly, this class is reused a few times. I'll keep the `std::string` from the function as it is very small to implement. I do like the `SetupInterface` approach, so I'll transition the casting over.

BruceWyxv replied:
Done. I also moved `MutableCoefficientsInterface` from utils/ to a directory that better represents its assignment in the context of functional expansions.

andrsd commented on modules/functional_expansion_tools/include/auxkernels/FunctionSeriesToAux.h:
```diff
@@ -0,0 +1,37 @@
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
```
You will want to use the new header (it is currently WIP, see #10468):
```
// This file is part of the MOOSE framework
// https://www.mooseframework.org
//
// All rights reserved, see COPYRIGHT for full restrictions
// https://github.com/idaholab/moose/blob/master/COPYRIGHT
//
// Licensed under LGPL 2.1, please see LICENSE for details
// https://www.gnu.org/licenses/lgpl-2.1.html
```
And some file are missing it

BruceWyxv replied:
Done

andrsd commented on modules/functional_expansion_tools/include/bcs/FEFluxBC.h:
```diff
@@ -0,0 +1,34 @@
+template <>
+InputParameters validParams<FEFluxBC>();
+
+/// Defines an FE-based BC that strongly encourages the gradients to match
```
Class description and methods use:
```
/**
 * Comment
 */
```

Member variables use:
```
/// comment
```

BruceWyxv replied:
Good to know.

I used the triple slash for single-line and the stars for multi-line. It will take a while, but I'll get the refactoring done for this.

andrsd commented on modules/functional_expansion_tools/src/bcs/FEValuePenaltyBC.C:
```diff
@@ -0,0 +1,45 @@
+  : FunctionPenaltyDirichletBC(parameters)
+{
+  /*
+   * The _func member was made private in FunctionPenaltyDirichletBC, so manually acquire the
```
Change the access to `_func` to protected in `FunctionPenaltyDirichletBC`

BruceWyxv replied:
Fine by me. I'll do this in a separate PR.

BruceWyxv replied:
PR submitted, #10581

Commit:
Migrated to new header

Added header to some files that were missing it

Commit:
Updated comments/documentation

/** ... */ style for class descriptions and methods
/// ... style for member variables

Commit:
Added formatted coefficient table getter to MutableCoefficientsInterface

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/148095/) on d126798 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10579/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10579/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7a18e52d8770a53d23ff4b961d11e9f0fdcc0dc6`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/148111/) on a1120ed wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10579/site)

This comment will be updated on new commits.

Commit:
Organized files by task instead of dumping in the "utils" directory

Commit:
Casting replacement in MutableCoefficientsInterface

Replaced the sideways `dynamic_cast<>()` to a `MooseObject` by requiring a pointer to the fully-qualified `MooseObject` in the constructor, then grabbing the reference to the `_console` instance. This is used directly to print the coefficients table to the console when requested.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/148329/) on b49f41f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10579/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10579/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7a18e52d8770a53d23ff4b961d11e9f0fdcc0dc6`

brianmoose commented:
@BruceWyxv An invitation has been sent to you to join the "MOOSE Users" group. Once accepted, your PRs will get automatically get tested and won't need a developer to activate.

Commit:
Added required documentation files

BruceWyxv commented:
@permcody I've got some time this afternoon to come over if that works?

BruceWyxv commented:
I'm not sure how to fix the failure of the Documentation check. The "Test python" step is failing with errors that don't seem to be related to any of the code in this PR. Any suggestions?

Commit:
Updated FEValuePenaltyBC

Changed FEValuePenaltyBC to use the recently protected `_func` in FunctionPenaltyDirichletBC.

brianmoose commented:
@BruceWyxv The "Test python" failures were not related to this PR and should be fixed now.  I restarted the tests, and they should pass now.

permcody commented:
@BruceWyxv - We are very busy this week with visitors. Can we push off your visit until next week?

BruceWyxv commented:
@permcody - Works for me. How does Monday or Tuesday look?

permcody commented:
Either is fine. Monday anytime or Tuesday in the afternoon. Let me know
what works for you and I'll make sure the rest of the team (that is
interested) can join.

On Wed, Jan 24, 2018 at 2:30 PM BruceWyxv <notifications@github.com> wrote:

> @permcody <https://github.com/permcody> - Works for me. How does Monday
> or Tuesday look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10579#issuecomment-360279952>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIDZPVlmFVuq1sV_HJnePhDe53m9Qks5tN6ELgaJpZM4Ror3w>
> .
>

BruceWyxv commented:
Monday would probably be best. I'll plan on coming by around 9 if that
works?

Is there a whiteboard that we can use to jot down ideas or diagrams?

On Wed, Jan 24, 2018 at 3:09 PM, Cody Permann <notifications@github.com>
wrote:

> Either is fine. Monday anytime or Tuesday in the afternoon. Let me know
> what works for you and I'll make sure the rest of the team (that is
> interested) can join.
>
> On Wed, Jan 24, 2018 at 2:30 PM BruceWyxv <notifications@github.com>
> wrote:
>
> > @permcody <https://github.com/permcody> - Works for me. How does Monday
> > or Tuesday look?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/10579#issuecomment-360279952>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AC5XIDZPVlmFVuq1sV_
> HJnePhDe53m9Qks5tN6ELgaJpZM4Ror3w>
> > .
>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10579#issuecomment-360290554>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHt6LSmSJyy7RcuyS_QF_C9r4aSqeeUYks5tN6omgaJpZM4Ror3w>
> .
>

permcody commented:
Yes, and yes, we'll see you around 9

On Wed, Jan 24, 2018 at 3:14 PM BruceWyxv <notifications@github.com> wrote:

> Monday would probably be best. I'll plan on coming by around 9 if that
> works?
>
> Is there a whiteboard that we can use to jot down ideas or diagrams?
>
> On Wed, Jan 24, 2018 at 3:09 PM, Cody Permann <notifications@github.com>
> wrote:
>
> > Either is fine. Monday anytime or Tuesday in the afternoon. Let me know
> > what works for you and I'll make sure the rest of the team (that is
> > interested) can join.
> >
> > On Wed, Jan 24, 2018 at 2:30 PM BruceWyxv <notifications@github.com>
> > wrote:
> >
> > > @permcody <https://github.com/permcody> - Works for me. How does
> Monday
> > > or Tuesday look?
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/idaholab/moose/pull/10579#issuecomment-360279952>,
> > or mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-auth/AC5XIDZPVlmFVuq1sV_
> > HJnePhDe53m9Qks5tN6ELgaJpZM4Ror3w>
> > > .
> >
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/10579#issuecomment-360290554>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AHt6LSmSJyy7RcuyS_QF_C9r4aSqeeUYks5tN6omgaJpZM4Ror3w
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10579#issuecomment-360291935>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIIAhUstNJamwjg9GvExzoJkTsk1qks5tN6tcgaJpZM4Ror3w>
> .
>

Commit:
Comment/documentation changes only - no code alterations

Updated COPYRIGHT to match recent PR
Unified header file inclusion comments
Rebased on top of current upstream/devel

Commit:
Updates based on meeting with MOOSE team 2018-02-05

Refactored FE -> FX to remove ambiguity between "functional expansion" and "finite-element"
Made MutableCoefficientsInterface restartable - required making MCI a friend class of `Restartable` to avoid diamond inheritance via `Function` and `UserObject` subclasses.
Updated README files to highlight potiential issues when using `FunctionSeriesToAux`+`CoupledForce` approach

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/153115/) on 300881c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10579/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10579/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0304029221a0732d6d9bda07c2b3a506c9b79029`

BruceWyxv commented:
I just pushed the changes that were discussed in out meeting today. Everything is rebased onto MOOSE's devel branch.

One particular change that I'd like to highlight is the need to add `MutableCoefficientsInterface` as a friend to `Restartable`. This was done to prevent diamond inheritance, since `MutableCoefficientsInterface` is subclassed both by `Function` and `UserObject`. This allowed me to pass `this` from the subclasses as a `Restartable *`, which I used to call the protected member `declareRestartableData<...>()`.

permcody commented:
@BruceWyxv - Can you help us out with documentation? See the specific objects needing docs here:
https://www.moosebuild.org/job/153125/

The syntax for creating docs is Markdown and we ask that you write a little bit about your objects. More than you might put in a comment.
http://mooseframework.org/docs/moose_docs/site/utilities/moose_docs/index.html

Commit:
Updated documentation files

andrsd commented on modules/functional_expansion_tools/src/transfers/MultiAppMutableCoefficientsTransfer.C:
```diff
@@ -0,0 +1,231 @@
+      "Name of the MutableCoefficientsInterface-derived object in the MultiApp.");
+
+  params.addParam<std::string>(
+      "pretty_name", "MultiAppFXTransfer", "Typename by which this class should be identified.");
```
You will want to change this after you remove that typedef.

andrsd commented on modules/functional_expansion_tools/unit/Makefile:
```diff
@@ -0,0 +1,72 @@
+-include $(functional_expansion_tools_unit_deps)
+
+###############################################################################
+# Additional special case targets should be added here
```
Replace all TAB characters with spaces

mangerij commented:
So does this work for arbitrary expansions of gradients of the order parameter? Vectorial order parameters?

andrsd commented on modules/modules.mk:
```diff
@@ -3,49 +3,50 @@
 
 ifeq ($(PHASE_FIELD),yes)
-        TENSOR_MECHANICS          := yes
+        TENSOR_MECHANICS          	:= yes
```
Replace all TAB characters with spaces.

Commit:
More documentation

Added page for module
Added examples page
Used built-in BibTeX system
Added comment in documentation about displaced meshes

Commit:
Remove Captain Obvious comments

BruceWyxv commented on modules/functional_expansion_tools/include/functions/FunctionSeries.h:
```diff
@@ -0,0 +1,127 @@
+ * convolved function series. Its inheritance tree includes MutableCoefficientsInterface, which
+ * enables easy MultiApp transfers of coefficients.
+ */
+class FunctionSeries : public MutableCoefficientsFunctionInterface
```
`FunctionSeries` doesn't join together 2 MOOSE systems through inheritance. It only creates the functional expansion as a `Function` type object.

`FX...UserObject`-type classes require a `FunctionSeries` object and use it to integrate the projection of the functional series on the field being integrated, i.e. composition. These projection values are stored as coefficients in the interface `MutableCoefficientsInterface`.

You are correct in that both the `FX...UserObject` and `FunctionSeries` inherit from `MutableCoefficientsInterface`. From where I see it, this is not much different than the way that both `UserObject` and `Function` inherit from `MooseObject` (or `Restartable` itself). However, these coefficients are the key feature of functional expansions. The transfer `MultipAppFETransfer` uses this interface to move these coefficients between supported objects.

Please let me know if there is a better way to get the `_characteristics` and `_coefficients` arrays of `MutableCoefficientsInterface` to be restartable.

andrsd replied:
I guess, I got tangled up in the diagrams from doxygen, but the way you had to use the `Restartable` was not natural to MOOSE systems and seems wrong. Let me go over that again.

BruceWyxv replied:
Another possible option would be to have all the subobjects create the restartable arrays themselves, passing the references as arguments to the `MutableCoefficientsInterface` constructor.

BruceWyxv commented on modules/functional_expansion_tools/include/functions/FunctionSeries.h:
```diff
@@ -0,0 +1,127 @@
+   * issues reading a list of integers in as `std::size_t` (unsigned long), so this workaround is
+   * required in order to set `_orders` in the constructor initializer list.
+   */
+  static std::vector<std::size_t> convertOrders(const std::vector<unsigned int> orders);
```
Nice catch, got it.

BruceWyxv commented on modules/functional_expansion_tools/include/series/FunctionalBasisInterface.h:
```diff
@@ -0,0 +1,180 @@
+
+// Shortened typename
+class FunctionalBasisInterface;
+typedef FunctionalBasisInterface FBI;
```
Meaning `FBI` isn't used or it shouldn't be used?

andrsd replied:
Meaning no aliases via `typedef`s, so replace all `FBI` with `FunctionalBasisInterface` and drop this `typedef`.

BruceWyxv commented on modules/functional_expansion_tools/include/series/FunctionalBasisInterface.h:
```diff
@@ -0,0 +1,180 @@
+/**
+ * Passkey for ...SeriesBasisInterface
+ */
+class SBIKey
```
I realized it isn't needed. I changed it so that `SingleSeriesBasisInterface` and `CompositeSeriesBasisInterface` both declare `clearBasisEvaluation(...)` as private.

BruceWyxv commented on modules/functional_expansion_tools/include/transfers/MultiAppMutableCoefficientsTransfer.h:
```diff
@@ -0,0 +1,87 @@
+};
+
+/// Typedef for the functional_expansion_tools module
+typedef MultiAppMutableCoefficientsTransfer MultiAppFXTransfer;
```
I like MultiAppFXTransfer better, using that everywhere...

andrsd replied:
Good choice 👍

BruceWyxv commented on modules/functional_expansion_tools/include/userobject/FXIntegralBaseUserObject.h:
```diff
@@ -0,0 +1,246 @@
+#include "MutableCoefficientsInterface.h"
+
+/**
+ * Class declaration for parameters - we cannot use templated types in validParams<>()
```
I tried something similar. At least it didn't work since I'm providing template parameter, not an actual type. Maybe I did it wrong?

`FXIntegralBaseUserObject` is a host class that provides a policy-based design. This base object generically allows a functional expansion to be derived from any `UserObject` that contains `computeQpIntegral()` as the policy class.

The same base `InputParameters` are required, no matter what policy is provided (although each policy may need to add some more like `FXBoundaryFluxUserObject`). I'd really hate to copy-paste the parameters for each policy. This is the reason that I configured an empty parameter class. I'm very open to a better solution.

BruceWyxv commented on modules/functional_expansion_tools/include/userobject/FXIntegralBaseUserObject.h:
```diff
@@ -0,0 +1,246 @@
+  virtual void threadJoin(const UserObject & sibling) final;
+
+  /// Keep the expansion coefficients after each solve
+  const bool _keep_history;
```
Nope, no need. It's now `protected`.

BruceWyxv commented on modules/functional_expansion_tools/include/utils/Hashing.h:
```diff
@@ -0,0 +1,93 @@
+inline HashValue
+hashCombine(const libMesh::Point & point)
+{
+  HashValue value = 3;
```
Does something need changed?

andrsd replied:
A comment that explains why 3. Why not 33? Or 333? :-)

BruceWyxv replied:
Or 42? :-)

I put in a comment. It's just a seed value, and provides entropy against the other seed value.

andrsd replied:
Also, this would be much better:
```
HashValue seed = 3;
```
the word `value` is just too general. Moreover, notice that here you use 3 and pass 3 parameters and below, you use 4 and pass 4 parameters.

BruceWyxv replied:
Got it `seed` it is.

I didn't even notice the coincidence with the number of arguments.

BruceWyxv commented on modules/functional_expansion_tools/src/functions/FunctionSeries.C:
```diff
@@ -0,0 +1,269 @@
+   */
+  MooseEnum SeriesTypes("Cartesian CylindricalDuo");
+  MooseEnum SingleSeriesTypes1D("Legendre");
+  MooseEnum SingleSeriesTypes2D("Zernike");
```
Whoops, fixed.

BruceWyxv commented on modules/functional_expansion_tools/src/functions/FunctionSeries.C:
```diff
@@ -0,0 +1,269 @@
+
+    types.push_back(_disc);
+    _series_type = new CylindricalDuo(domains, _orders, types);
+  }
```
Where'd all my virtual destructors go? I must have been too aggressive when removing the empty ones and accidentally removed the ones I actually needed. In practice I always code `new` and `delete` at the same time to prevent this from happening.

I've added it back in. There really is not a good way to do this without raw memory allocation since `FunctionSeries` is working with generic `CompositeSeriesBasisInterface` instances, of which `Cartiesian` and `CylindricalDuo` are the current options. Maybe more to come in the future?

BruceWyxv replied:
I think I misunderstood. Are you referring to using `libmesh_make_unique<T>(...)` or similar instead?

andrsd replied:
`std::shared_ptr<T>`

andrsd replied:
But, if you think no one else should get their dirty hands on your pointers, then `unique_ptr` is good.

BruceWyxv replied:
No touchy

permcody replied:
> @permcody might bitch about this

You bet I will. I don't want to see the keyword "new" in this new module 😄 .

BruceWyxv replied:
:relaxed: no more `new` anywhere after 3f578e9. Only `std::unique_ptr`.

For completeness the word "new" does show up in a few comments and variable names. :roll_eyes:

BruceWyxv commented on modules/functional_expansion_tools/src/functions/MemoizedFunctionInterface.C:
```diff
@@ -0,0 +1,84 @@
+  {
+    // Start the cache over if we are at a new time step
+    if (_respect_time && time != _current_time)
+      _current_time = time, invalidateCache();
```
Despise them! :-D

JK, I think I was showing someone the comma operator and then left it as is. I'm changing it back: I'm not attached to it and the blocks will increase legibility.

BruceWyxv commented on modules/functional_expansion_tools/src/series/Cartesian.C:
```diff
@@ -0,0 +1,50 @@
+  // Initialize the pointers to each of the single series
+  for (std::size_t i = 0; i < _series_types.size(); ++i)
+    if (_series_types[i] == "Legendre")
+      _series.push_back(new Legendre({domains[i]}, {orders[i]}));
```
These are all now deleted in `CompositeSeriesBasisInterface` now that I have added back in the destructors.

BruceWyxv replied:
I swapped to using `std::unique_ptr` + `libmesh_make_unique` so there are no more `new` or `delete` calls in the module code that need managed.

BruceWyxv commented on modules/functional_expansion_tools/src/series/CylindricalDuo.C:
```diff
@@ -0,0 +1,54 @@
+
+CylindricalDuo::CylindricalDuo() : CompositeSeriesBasisInterface()
+{
+  // Nothing here, we only needed to call the parent constructor
```
LOL He's everywhere. Sorry about that, pulling all of these out.

BruceWyxv commented on modules/functional_expansion_tools/src/series/CompositeSeriesBasisInterface.C:
```diff
@@ -0,0 +1,299 @@
+  : FunctionalBasisInterface(), _series_types(series_types)
+{
+  if (orders.size() != _series_types.size())
+    mooseError("CSBI: Incorrect number of 'orders' specified for 'FunctionSeries'! Check that "
```
Good idea, but I'll pass in a reference to the string instead. This class itself doesn't sub/super class any `MooseObject`. At least it will show which object instance using `CompositeSeriesBasisInterface` is causing the problems.

Commit:
Changes based on comments by @andrsd (2018-02-06)
Renamed MultiAppMutableCoefficientsTransfer to MultiAppFXTransfer and removed typedef
Migrated to using `unique_ptr` for SSBI and CSBI pointers
Other minor fixes (spelling, whitespaces, comments, member visibility)

BruceWyxv commented:
I'm not sure I'm following fully. Can you please explain?

On Tue, Feb 6, 2018 at 11:05 AM, John <notifications@github.com> wrote:

> So does this work for arbitrary expansions of gradients of the order
> parameter? Vectorial order parameters?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10579#issuecomment-363511718>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHt6LX5F11AC08jiy0eUTDDsBEih6Cx0ks5tSJR6gaJpZM4Ror3w>
> .
>

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/153761/) on 519bba3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10579/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10579/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0304029221a0732d6d9bda07c2b3a506c9b79029`

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/153773/) on 12d7680 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.org/docs/PRs/10579/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.org/docs/PRs/10579/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0304029221a0732d6d9bda07c2b3a506c9b79029`

BruceWyxv commented:
I can't see the results of the Unity build check. Is there an issue I need to resolve?

BruceWyxv commented:
@permcody @friedmud Does this generalized diagram look better?
[FX_Diagram.pdf](https://github.com/idaholab/moose/files/1704152/FX_Diagram.pdf)

friedmud commented on modules/functional_expansion_tools/src/functions/MemoizedFunctionInterface.C:
```diff
@@ -0,0 +1,85 @@
+}
+
+Real
+MemoizedFunctionInterface::value(Real time, const Point & point)
```
This is not threadsafe.  You need to use a "doubly checked lock" with a C++11 memory fence.

The one called: "Using C++11 Acquire and Release Fences" here: http://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/

BruceWyxv replied:
Would it be possible for multiple threads to be evaluating the same `Point`? If so, I will probably need to lock the entire function body. That may incur a measurable performance hit.

Otherwise, I'm thinking it will be enough to lock just the writing operations.

friedmud reviewed (changes requested):
In addition to the threading issue: I'm also seeing many failed tests on my system using your branch:

```bash
Final Test Results:
--------------------------------------------------------------------------------------------------------------
test:errors.Cartesian_missing_series ...................................................................... OK
test:errors.invalid_bounds_length ......................................................................... OK
test:errors.CylindricalDuo_multiple_series ................................................................ OK
test:errors.CylindricalDuo_missing_disc_series ............................................................ OK
test:errors.FunctionSeriesToAux_bad_function .............................................................. OK
test:errors.invalid_order ................................................................................. OK
test:errors.MultiAppFXTransfer_bad_function_series ........................................................ OK
test:errors.CylindricalDuo_missing_axis_series ............................................................ OK
test:errors.MultiAppFXTransfer_incompatible_orders ........................................................ OK
test:errors.MultiAppFXTransfer_bad_user_object ............................................................ OK
test:errors.MultiAppFXTransfer_missing_sub ................................................................ OK
test:errors.MultiAppFXTransfer_missing_local .............................................................. OK
test:standard_use.neglect_invalid_enum ........................................................ FAILED (CRASH)
test:standard_use.incompatible_dimensions ..................................................... FAILED (CRASH)
test:standard_use.volume_coupling ............................................................. FAILED (CRASH)
test:standard_use.print_coefficients .......................................................... FAILED (CRASH)
test:standard_use.interface_coupling .......................................... [min_threads=2] FAILED (CRASH)
--------------------------------------------------------------------------------------------------------------
Ran 17 tests in 2.2 seconds
12 passed, 0 skipped, 0 pending, 5 FAILED
```

They all just crash... some say "illegal instruction" like this:

```bash
test:standard_use.interface_coupling: Postprocessor Values:
test:standard_use.interface_coupling: +----------------+-------------------------+-------------------+----------------+
test:standard_use.interface_coupling: | time           | average_interface_value | picard_iterations | total_flux     |
test:standard_use.interface_coupling: +----------------+-------------------------+-------------------+----------------+
test:standard_use.interface_coupling: |   0.000000e+00 |            0.000000e+00 |      1.000000e+00 |   0.000000e+00 |
test:standard_use.interface_coupling: |   1.000000e+00 |            5.033475e+00 |      8.000000e+00 |   3.364801e+01 |
test:standard_use.interface_coupling: |   2.000000e+00 |            6.865999e+00 |      8.000000e+00 |   3.585695e+01 |
test:standard_use.interface_coupling: |   3.000000e+00 |            8.364726e+00 |      7.000000e+00 |   3.610225e+01 |
test:standard_use.interface_coupling: |   4.000000e+00 |            9.744072e+00 |      7.000000e+00 |   3.615766e+01 |
test:standard_use.interface_coupling: +----------------+-------------------------+-------------------+----------------+
test:standard_use.interface_coupling: 
test:standard_use.interface_coupling: 
test:standard_use.interface_coupling: Advancing MultiApps
test:standard_use.interface_coupling: Waiting For Other Processors To Finish
test:standard_use.interface_coupling: Jobs complete: 1/1
test:standard_use.interface_coupling: Finished Advancing MultiApps
test:standard_use.interface_coupling: 
test:standard_use.interface_coupling: 
test:standard_use.interface_coupling: FXTransferApp0: 
test:standard_use.interface_coupling: 
test:standard_use.interface_coupling: ===================================================================================
test:standard_use.interface_coupling: =   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
test:standard_use.interface_coupling: =   PID 98947 RUNNING AT Dereks-Mac-Pro.inl.gov
test:standard_use.interface_coupling: =   EXIT CODE: 4
test:standard_use.interface_coupling: =   CLEANING UP REMAINING PROCESSES
test:standard_use.interface_coupling: =   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
test:standard_use.interface_coupling: ===================================================================================
test:standard_use.interface_coupling: YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Illegal instruction: 4 (signal 4)
test:standard_use.interface_coupling: This typically refers to a problem with your application.
test:standard_use.interface_coupling: Please see the FAQ page for debugging suggestions
test:standard_use.interface_coupling: 
test:standard_use.interface_coupling: ################################################################################
test:standard_use.interface_coupling: Tester failed, reason: CRASH
```

BruceWyxv commented:
@friedmud Now we enter the fun zone of "Well, it runs fine on my machine..."

Is there any additional output when testing with `--dbg`? Have you tried running with `gdb` to see where it crashes?

permcody commented:
@BruceWyxv - Before you start debugging you would normally make sure that you are Valgrind clean. I'd normally just add the Valgrind target to this PR and you'd get back some useful results. 

However, we currently have a bug in libMesh that is causing MOOSE to fail the Valgrind target. We are working on that and may merge in a new libMesh today which will fix this. That'll help us track down any memory corruption or leaks that might point to the illegal instruction problem here. If you want to get ahead, you can grab a later version of libMesh, recompile everything and then Valgrind on your tests.

```
$ ./run_tests --re=<pattern_for_your_tests> --valgrind -j 8
```

friedmud commented:
It is possible for functions to be evaluated at the same point... but even beyond that... the way that you've implemented `FunctionSeries::evaluateValue()` is very non-threadsafe.

You need to work to make `evaluateValue()` be `const` (and actually _be_ const).  Doing things like `setLocation()` or any other manipulation of class data should not be allowed in `evaluateValue()`.

Once you've done that you should be able to use a doubly checked lock with memory fences around the hash table lookup.

The code would check to see if that hash entry already exists.  If it does, then just read out of it.  If it doesn't... then lock and create that entry then release the lock.

If you've never run into this before... the reason for the memory fences is that it's possible for other processors to "see" that the entry exists... before the value of it is actually written into that memory position.  That causes a race condition.  The memory fences prevent that from being possible (the position in the hash won't be available to be read from until the value is actually written to it).

friedmud commented:
As for your memory corruption thing... I have no idea.  Maybe someone else can try and see if they get the same thing I did?  @permcody ?

friedmud commented:
It looks like the illegal instruction is coming during tear-down.  Here is a stack trace:

```
rocess 35922 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
    frame #0: 0x00000001001535e8 libfunctional_expansion_tools-dbg.0.dylib`CompositeSeriesBasisInterface::~CompositeSeriesBasisInterface(this=0x000000010d894e00) at CompositeSeriesBasisInterface.h:22
   19  	 * This class is the basis for constructing a composite---or convolved---functional series by
   20  	 * combining multiple other series together. Nonseparability is currently assumed.
   21  	 */
-> 22  	class CompositeSeriesBasisInterface : public FunctionalBasisInterface
   23  	{
   24  	public:
   25  	  CompositeSeriesBasisInterface(const std::string & who_is_using_me);
Target 0: (functional_expansion_tools-dbg) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
  * frame #0: 0x00000001001535e8 libfunctional_expansion_tools-dbg.0.dylib`CompositeSeriesBasisInterface::~CompositeSeriesBasisInterface(this=0x000000010d894e00) at CompositeSeriesBasisInterface.h:22
    frame #1: 0x0000000100153465 libfunctional_expansion_tools-dbg.0.dylib`FunctionSeries::~FunctionSeries() [inlined] std::__1::default_delete<CompositeSeriesBasisInterface>::operator(this=0x000000010d894910, __ptr=0x000000010d894e00)(CompositeSeriesBasisInterface*) const at memory:2233
    frame #2: 0x000000010015344a libfunctional_expansion_tools-dbg.0.dylib`FunctionSeries::~FunctionSeries() [inlined] std::__1::unique_ptr<CompositeSeriesBasisInterface, std::__1::default_delete<CompositeSeriesBasisInterface> >::reset(this=0x000000010d894910, __p=0x0000000000000000) at memory:2546
    frame #3: 0x00000001001533e5 libfunctional_expansion_tools-dbg.0.dylib`FunctionSeries::~FunctionSeries() [inlined] std::__1::unique_ptr<CompositeSeriesBasisInterface, std::__1::default_delete<CompositeSeriesBasisInterface> >::~unique_ptr(this=0x000000010d894910) at memory:2500
    frame #4: 0x00000001001533e5 libfunctional_expansion_tools-dbg.0.dylib`FunctionSeries::~FunctionSeries() [inlined] std::__1::unique_ptr<CompositeSeriesBasisInterface, std::__1::default_delete<CompositeSeriesBasisInterface> >::~unique_ptr(this=0x000000010d894910) at memory:2500
    frame #5: 0x00000001001533e5 libfunctional_expansion_tools-dbg.0.dylib`FunctionSeries::~FunctionSeries(this=0x000000010d894628) at FunctionSeries.h:26
    frame #6: 0x0000000100153325 libfunctional_expansion_tools-dbg.0.dylib`FunctionSeries::~FunctionSeries(this=0x000000010d894628) at FunctionSeries.h:26
    frame #7: 0x00000001001531a1 libfunctional_expansion_tools-dbg.0.dylib`std::__1::__shared_ptr_emplace<FunctionSeries, std::__1::allocator<FunctionSeries> >::__on_zero_shared(this=0x000000010d894610) at memory:3604
    frame #8: 0x00000001008f5d31 libmoose-dbg.0.dylib`std::__1::shared_ptr<Function>::~shared_ptr() [inlined] std::__1::__shared_count::__release_shared(this=0x000000010d894610) at memory:3438
    frame #9: 0x00000001008f5ce7 libmoose-dbg.0.dylib`std::__1::shared_ptr<Function>::~shared_ptr() [inlined] std::__1::__shared_weak_count::__release_shared(this=0x000000010d894610) at memory:3480
    frame #10: 0x00000001008f5ce7 libmoose-dbg.0.dylib`std::__1::shared_ptr<Function>::~shared_ptr(this=0x000000010d8a4700) at memory:4406
    frame #11: 0x00000001006a5f85 libmoose-dbg.0.dylib`std::__1::shared_ptr<Function>::~shared_ptr(this=0x000000010d8a4700) at memory:4404
    frame #12: 0x000000010076c539 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~__vector_base() [inlined] std::__1::allocator<std::__1::shared_ptr<Function> >::destroy(this=0x000000010d8503e0, __p=0x000000010d8a4700) at memory:1814
    frame #13: 0x000000010076c530 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~__vector_base() [inlined] void std::__1::allocator_traits<std::__1::allocator<std::__1::shared_ptr<Function> > >::__destroy<std::__1::shared_ptr<Function> >(__a=0x000000010d8503e0, __p=0x000000010d8a4700) at memory:1682
    frame #14: 0x000000010076c520 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~__vector_base() [inlined] void std::__1::allocator_traits<std::__1::allocator<std::__1::shared_ptr<Function> > >::destroy<std::__1::shared_ptr<Function> >(__a=0x000000010d8503e0, __p=0x000000010d8a4700) at memory:1550
    frame #15: 0x000000010076c50a libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~__vector_base() [inlined] std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::__destruct_at_end(this=0x000000010d8503d0, __new_last=0x000000010d8a4700) at vector:413
    frame #16: 0x000000010076c487 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~__vector_base() [inlined] std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::clear(this=0x000000010d8503d0) at vector:356
    frame #17: 0x000000010076c46f libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~__vector_base(this=0x000000010d8503d0) at vector:441
    frame #18: 0x000000010076c425 libmoose-dbg.0.dylib`std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~vector(this=0x000000010d8503d0 size=1) at iterator:1407
    frame #19: 0x000000010076c405 libmoose-dbg.0.dylib`std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >::~vector(this=0x000000010d8503d0 size=1) at iterator:1407
    frame #20: 0x000000010076cc59 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~__vector_base() [inlined] std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > >::destroy(this=0x000000010f002ff8, __p=0x000000010d8503d0 size=1) at memory:1814
    frame #21: 0x000000010076cc50 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~__vector_base() [inlined] void std::__1::allocator_traits<std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::__destroy<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > >(__a=0x000000010f002ff8, __p=0x000000010d8503d0 size=1) at memory:1682
    frame #22: 0x000000010076cc40 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~__vector_base() [inlined] void std::__1::allocator_traits<std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::destroy<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > >(__a=0x000000010f002ff8, __p=0x000000010d8503d0 size=1) at memory:1550
    frame #23: 0x000000010076cc2a libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~__vector_base() [inlined] std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::__destruct_at_end(this=0x000000010f002fe8, __new_last=0x000000010d8503d0 size=1) at vector:413
    frame #24: 0x000000010076cba7 libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~__vector_base() [inlined] std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::clear(this=0x000000010f002fe8) at vector:356
    frame #25: 0x000000010076cb8f libmoose-dbg.0.dylib`std::__1::__vector_base<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~__vector_base(this=0x000000010f002fe8) at vector:441
    frame #26: 0x000000010076cb45 libmoose-dbg.0.dylib`std::__1::vector<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~vector(this=0x000000010f002fe8 size=1) at iterator:1407
    frame #27: 0x000000010076b5d5 libmoose-dbg.0.dylib`std::__1::vector<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > >, std::__1::allocator<std::__1::vector<std::__1::shared_ptr<Function>, std::__1::allocator<std::__1::shared_ptr<Function> > > > >::~vector(this=0x000000010f002fe8 size=1) at iterator:1407
    frame #28: 0x000000010076b573 libmoose-dbg.0.dylib`MooseObjectWarehouseBase<Function>::~MooseObjectWarehouseBase(this=0x000000010f002fd8) at MooseObjectWarehouseBase.h:218
    frame #29: 0x000000010076b4e5 libmoose-dbg.0.dylib`MooseObjectWarehouse<Function>::~MooseObjectWarehouse(this=0x000000010f002fd8) at MooseObjectWarehouse.h:25
    frame #30: 0x000000010067f0c5 libmoose-dbg.0.dylib`MooseObjectWarehouse<Function>::~MooseObjectWarehouse(this=0x000000010f002fd8) at MooseObjectWarehouse.h:25
    frame #31: 0x00000001006820a4 libmoose-dbg.0.dylib`FEProblemBase::~FEProblemBase(this=0x000000010f002c18) at FEProblemBase.C:327
```

friedmud commented:
Ok - I was wrong about threading.

It turns out that we _do_ create one Function instance per thread.  That means that your code is ok (kind of) and you don't need the doubly checked lock or anything.

It also means that your cache is going to be less effective when running with threads because each thread-instanced Function will maintain its own cache and the threads aren't guaranteed to go over the same elements (actually - they are with pthreads - but not with TBB).

BruceWyxv commented:
But I should probably still revise the `evaluateValue()` to ensure true `const` behavior, yes?

I can deal with a less-effective cache for now. The objective is to avoid evaluating the entire functional expansion at every point for all single linear/non-linear solves.

friedmud commented:
It's good enough like it is.  Sorry for the run-around!

We're actually making some changes to the threading model (probably tomorrow) that will actually make your cache fairly effective (the same threads should run over the same elements).

Now all that's left is you need to fixup whatever is going wrong with your destructors!

Commit:
Added virtual dtors to SSBI and CSBI

Also disabled move and copy operations

BruceWyxv commented:
@friedmud No worries, thanks for letting me know.

I cannot replicate the tear-down issue on my system. I stepped my libMesh back to 09776be, but valgrind is still reporting only a libMesh error. I can't go to an earlier update because of missing functions, unless there is a specific non-MOOSE-update commit that is known to be good?

I'm guessing the compiler is doing some magic behind the scenes to save my bacon.

I've pushed some changes with with explicit virtual destructors in CSBI and SSBI. Can you please test to see if this fixes the issues on your system?

BruceWyxv commented:
Should I be merging/rebasing from upstream/devel throughout this process?

friedmud commented:
Ok - it turns out that the problem I was seeing is because your branch from a fairly old version of MOOSE and I'm using an up to date libMesh.  Once I rebased your branch it worked fine.

So: In this instance... it could be a good idea for you to rebase this branch - just to make reviewing it that much easier.

However: I have now uncovered a new problem.  You `#define MAX_DIRECT_CALCULATION` in two different `.C`... that fails when doing Unity builds.  Please give them unique names in each `.C` file.

Commit:
Updated `Legendre.C` and `Zernike.C`

Fixed compiler warnings
Made `MAX_DIRECT_CALCULATION` macros unique
Added `libmesh_fallthrough();` to intentional case fallthroughs

friedmud commented:
The Unity build didn’t fail because of anything you did though.
On Thu, Feb 8, 2018 at 12:28 PM BruceWyxv <notifications@github.com> wrote:

> The branch was rebased.
>
> The #define MAX_DIRECT_CALCULATION macros are now unique, but it appears
> that the Unity build still failed. What else needs revised?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10579#issuecomment-364221718>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMbQhUexhkmTCGmjZhjiuXYGrJ1UZks5tS0qugaJpZM4Ror3w>
> .
>

BruceWyxv commented:
@andrsd @friedmud Have I addressed all the comments and required changes? Please let me know what remains so I can fix those as well.

friedmud commented on framework/include/restart/Restartable.h:
```diff
@@ -211,6 +211,9 @@ class Restartable
   friend class TransientMultiApp;
+
+  /// For access to declareRestartableData() while avoiding diamond inheritance
+  friend class MutableCoefficientsInterface;
```
I would rather move `declareRestartableData()` to being public.  What do you think @permcody ?

permcody replied:
Yes, but I'm also OK, if we put this in now. I looked at this briefly this morning and it'll be easy to remove this after the fact. I'm willing to fix this up myself. Valgrind tests have passed.

friedmud commented on modules/functional_expansion_tools/src/functions/FunctionSeries.C:
```diff
@@ -0,0 +1,269 @@
+FunctionSeries &
+FunctionSeries::checkAndConvertFunction(Function & function, const std::string & name)
+{
+  if (!dynamic_cast<FunctionSeries *>(&function))
```
Use `libmesh_dynamic_cast` instead.  It compiles to a `static_cast` in optimized mode.

BruceWyxv replied:
Do you mean `libmesh_cast_ptr`?

I looks like it already performs an error check and report internally, so that simplifies this function to a one-liner (albeit with a little less contextual information)

BruceWyxv replied:
Looking through the use cases, the whole idea is to report to the user if they provided a `Function` object that is not a `FunctionSeries` type. It has nothing to do with ensuring the casting I'm doing during development is safe.

In other words, I need the `dynamic_cast` in all operational modes.

However, when looking as this I realized that many of my reconstruction objects perform their own checks. I will push an update soon that universally uses this static function instead.

Commit:
Unified use of FunctionSeries:checkAndConvertFunction()

Also added tests to verify checks

friedmud commented:
This all looks good - great job!

BruceWyxv commented:
Thanks everyone for your help, comments, and time with this initial development.

Once again, this has been a fantastic experience for me.
